Core/GameEngine/Source/GameNetwork/GameInfo.cpp

```diff
@@ -892,7 +895,130 @@ Bool GameInfo::isSandbox(void)
+	{
+		if (lhs.Length == rhs.Length)
+			return lhs.Index < rhs.Index;
+		return lhs.Length < rhs.Length;
+	}
+	friend bool operator>(const LengthIndexPair& lhs, const LengthIndexPair& rhs) { return rhs < lhs; }
+	friend bool operator<=(const LengthIndexPair& lhs, const LengthIndexPair& rhs) { return !(lhs > rhs); }
+	friend bool operator>=(const LengthIndexPair& lhs, const LengthIndexPair& rhs) { return !(lhs < rhs); }
+};
+
+static AsciiStringVec BuildPlayerNames(const GameInfo& game)
+{
+	AsciiStringVec playerNames;
+	playerNames.resize(MAX_SLOTS);
+
+	for (Int i = 0; i < MAX_SLOTS; ++i)
+	{
+		const GameSlot* slot = game.getConstSlot(i);
+		if (slot->isHuman())
+		{
+			playerNames[i] = WideCharStringToMultiByte(slot->getName().str()).c_str();
+		}
+	}
+
+	return playerNames;
+}
+
+static Int SumLength(Int sum, const LengthIndexPair& l)
+{
+	return sum + l.Length;
+}
+
+static Bool TruncatePlayerNames(AsciiStringVec& playerNames, Int truncateAmount)
+{
+	// wont truncate any name to below this length
+	constexpr const Int MinimumNameLength = 1;
+
+	// make length+index pairs for the player names
+	std::vector<LengthIndexPair> lengthIndex;
+	lengthIndex.resize(playerNames.size());
+	for (size_t pi = 0; pi < playerNames.size(); ++pi)
+	{
+		Int playerNameLength = playerNames[pi].getLength();
+		lengthIndex[pi].Length = std::max(0, playerNameLength - MinimumNameLength);
+		lengthIndex[pi].Index = pi;
+	}
+
+	Int remainingNamesLength = std::accumulate(lengthIndex.begin(), lengthIndex.end(), 0, SumLength);
+
+	if (truncateAmount > remainingNamesLength)
+	{
+		DEBUG_LOG(("TruncatePlayerNames - Requested to truncate %u chars from player names, but only %u were available for truncation.",
+			truncateAmount, remainingNamesLength));
+		return false;
+	}
+
+	// sort based on length in ascending order
+	std::sort(lengthIndex.begin(), lengthIndex.end());
+
+	Int truncateNameAmount = 0;
+	for (size_t i = 0; i < lengthIndex.size(); ++i)
+	{
+		// round avg name length up, which will penalize the final entry (longest name) as it will have to account for the roundings
+		Int avgNameLength = ((remainingNamesLength - truncateAmount) + (playerNames.size() - i - 1)) / (playerNames.size() - i);
+		remainingNamesLength -= lengthIndex[i].Length;
+		if (lengthIndex[i].Length <= avgNameLength)
+		{
+			continue;
+		}
+
+		// ensure a longer name is not truncated less than a previous, shorter name
+		truncateNameAmount = std::max(truncateNameAmount, lengthIndex[i].Length - avgNameLength);
+		if (i == lengthIndex.size() - 1)
+		{
+			// ensure we account for rounding errors when truncating the last, longest entry
+			truncateNameAmount = std::max(truncateAmount, truncateNameAmount);
+		}
+
+		// as the name is UTF-8, make sure we don't truncate part of a multibyte character
+		while (lengthIndex[i].Length - truncateNameAmount >=  MinimumNameLength
+			&& (playerNames[lengthIndex[i].Index][lengthIndex[i].Length - truncateNameAmount + 1] & 0xC0) == 0x80)
```

Comment:
style: instead of using complex `lengthIndex[i].Index` and `lengthIndex[i].Length` repeatedly and making it difficult to read, cache them at the start of the loop with new short names.